core/shared/src/main/scala/zio/ZQueue.scala

```diff
@@ -484,104 +483,118 @@ object ZQueue {
 
     val capacity: Int = queue.capacity
 
-    def offer(a: A): UIO[Boolean] = offerAll(List(a))
+    def offer(a: A): UIO[Boolean] =
+      UIO.effectSuspendTotal {
+        if (shutdownFlag.get) ZIO.interrupt
+        else {
+          val noRemaining =
+            unsafePollN(takers, 1) match {
```

Comment:
May be worth drilling down into `unsafePollN` and optimizing it. For example, right now it is building this `List` every time you call it.

Reply:
Will do, probably the same specialization can be done with offering a single element.